### ISSUE TITLE
YOLOv7 and DeepSORT | Setup and Documentation

### ISSUE THREAD

**Issue by chronomustard** [chronomustard/YOLOv7-DeepSORT-Thermal#2]
# YOLOv7 DeepSORT ReId Documentation

This documentation serves as a guide for developers on the integration of YOLOv7 and DeepSORT for Reidentification Task.
This documentation can also be accesed from this [github page](https://chronomustard.github.io/YOLOv7-DeepSORT-Thermal/).

## Released Versions

- **Pretrained-base v0.1:** The first release of the pretrained weight for thermal datasets.

## Architecture

This ReId algorithm is built with YOLO and DeepSORT on Python 3.7.16. The dependencies are listed inside the `requirements.txt` from the repo.

---

### Minimum Requirements

As of February 2024, a minimum GPU that supports CUDA version 11.6 is needed for this.

### Installing Conda

To ease the configuration of Python and its dependencies, we will use [Conda](https://docs.conda.io/projects/conda/en/latest/user-guide/install/).

**Important:** The virtual environment used can be downloaded [here](https://github.com/chronomustard/YOLOv7-DeepSORT-Thermal/tree/main/env).

### Configuring Tensorflow and PyTorch

To allow the use of Tensorflow and PyTorch, respective CUDA and Cudatoolkit is required. Please refer to these respective guides for [CUDA (Windows)](https://docs.nvidia.com/cuda/cuda-installation-guide-microsoft-windows/index.html), [CUDA (Linux)](https://docs.nvidia.com/cuda/cuda-installation-guide-linux/index.html), [Tensorflow](https://pytorch.org/get-started/previous-versions/), and [Pytorch](https://pytorch.org/get-started/previous-versions/). For reference, in this project, TensorFlow 1.15, PyTorch 1.13+cu116 with torchaudio and torchvision were used.

### Datasets and Pretrained Weights

A market-1501-based custom dataset compiled from ThermalDB, RegDB, market-1501, and IndoThermal is used. The pretrained weights can be accessed [here](#).

---

### Training on Custom Dataset

Guides for training YOLOv7 with a custom dataset can be found [here](https://blog.roboflow.com/yolov7-custom-dataset-training-tutorial/). A [notebook](#) in the repo can be used for training YOLO.

### Training Evaluation

Validation and evaluation produce several parameters. Evaluation parameters, including graphs, will be displayed after training.

### Standalone Testing and Inference

For standalone image detection, follow this [guide](https://learnopencv.com/yolov7-object-detection-paper-explanation-and-inference/). DeepSORT is required for object reidentification.

### Save the Weight for Later Reidentification Inference with DeepSORT

After training, YOLOv7 model weights will be in `runs/detect/trainX`. Choose the best result based on your preference.

---

### Environment Setup

Make sure the [conda environment](https://github.com/chronomustard/YOLOv7-DeepSORT-Thermal/tree/main/env) is downloaded. Create an environment using the provided `.yml` file.

### Dataset Setup

For training DeepSORT, the dataset needs to be cropped in the market-1501 format. Scripts are provided [here](https://github.com/chronomustard/YOLOv7-DeepSORT-Thermal/tree/main/deepSORT_preprocessing_script) for dataset formatting.

### Training the DeepSORT Tracker

To train, navigate to `DEEP_SORT/DEEP DIRECTORY` and run:

```bash
conda activate torch1131
train.py --data-dir="MARKET-1501-BASED DATASET"
```

### Training Evaluation

Graphs will be generated after training. Ensure loss and top1err approach 0.

### Save the DeepSORT Tracker Weight

After training, the DeepSORT model weight will be in `deep_sort/deep/checkpoint`.

---

### DeepSORT-YOLOv7 ReId Configuration

The program is provided [here](#). Download [this folder](https://drive.google.com/drive/folders/1lbdhDOpZ3475eekVQ2q2e7rsli2dffll?usp=sharing) and configure the DeepSORT checkpoint and YAML file.

Before inference, change the YOLOv7 weight path and video file path in the notebook.

### Reidentification Inference

After running the code snippet, the inference result will be inside `yolov7/runs/detect`.

---

## References

- [Conda User Guide](https://docs.conda.io/projects/conda/en/latest/user-guide/install/)
- [YOLOv7-DeepSORT-Thermal Environment](https://github.com/chronomustard/YOLOv7-DeepSORT-Thermal/tree/main/env)
- [CUDA Installation Guide for Microsoft Windows](https://docs.nvidia.com/cuda/cuda-installation-guide-microsoft-windows/index.html)
- [CUDA Installation Guide for Linux](https://docs.nvidia.com/cuda/cuda-installation-guide-linux/index.html)
- [Get Started with PyTorch - Previous Versions](https://pytorch.org/get-started/previous-versions/)
- [YOLOv7 Custom Dataset Training Tutorial](https://blog.roboflow.com/yolov7-custom-dataset-training-tutorial/)
- [YOLOv7-DeepSORT-Thermal Environment](https://github.com/chronomustard/YOLOv7-DeepSORT-Thermal/tree/main/env)
- [YOLOv7 Object Detection Paper Explanation and Inference](https://learnopencv.com/yolov7-object-detection-paper-explanation-and-inference/)
- [DeepSORT PyTorch](https://github.com/ZQPei/deep_sort_pytorch)
- [YOLOv7-DeepSORT-Thermal Environment](https://github.com/chronomustard/YOLOv7-DeepSORT-Thermal/tree/main/env)
- [Scalable Person Re-Identification: A Benchmark](https://openaccess.thecvf.com/content_iccv_2015/html/Zheng_Scalable_Person_Re-Identification_ICCV_2015_paper.html)
- [Market-1501 Dataset](https://www.kaggle.com/datasets/whurobin/market-1501)
- [YOLOv7-DeepSORT-Thermal Preprocessing Scripts](https://github.com/chronomustard/YOLOv7-DeepSORT-Thermal/tree/main/deepSORT_preprocessing_script)
- [YOLOv7-DeepSORT-Thermal](https://github.com/chronomustard/YOLOv7-DeepSORT-Thermal/tree/main)

**Comment by chronomustard** [comment]
If you have any problems, please don't hesitate to create another issue!
